### PR TITLE
ext_authz: support external authorization checks when a route contains a direct response entry

### PR DESCRIPTION
**Pull request by esmet**
Commit Message: ext_authz: support external authorization checks when a route contains a direct response entry
Additional Description:
Risk Level: low
Testing: TODO
Docs Changes:
Release Notes: ext_authz: fix a bug that skipped external authorization checks for routes that contained only a direct response.
Fixes #https://github.com/envoyproxy/envoy/issues/14984

### CODE REVIEW

**Comment by esmet** [comment]
@dio do you mind providing a quick first pass while I work on test(s) and addressing feedback on other PRs?

**Comment by dio** [comment]
Thank you for working on this. Yeah, I think this makes sense.

**Comment by esmet** [comment]
There are tests so I'm pulling it out of draft 👍

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 1]
```diff
@@ -376,8 +376,16 @@ void Filter::continueDecoding() {
...
 Filter::PerRouteFlags Filter::getPerRouteFlags(const Router::RouteConstSharedPtr& route) const {
-  if (route == nullptr || route->routeEntry() == nullptr) {
-    return PerRouteFlags{true /*skip_check_*/, false /*skip_request_body_buffering_*/};
+  if (route == nullptr) {
+    // If there's no route, skip authorization checks. This is an optimization to avoid work
+    // if we know the request won't be forwarded upstream anyway.
+    return PerRouteFlags::skipCheckFlags();
+  }
+
+  if (route->routeEntry() == nullptr && route->directResponseEntry() == nullptr) {
```
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:385]: Could you quickly scan in the docs, if we have the opportunity to inform this explicitly in the docs?

**Review by dio — CHANGES_REQUESTED** [review]
Thanks for working on this! Looks good overall. Asking for improving the docs and an opportunity for optimization.

File: source/extensions/filters/http/ext_authz/ext_authz.cc (hunk 1 above)
Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:385]: @dio Good callout. I read the docs. I think we're actually going to skip authz checks for redirect actions still. Maybe this feature would be more complete if we:
- Documented that only routes that proxy requests upstream will be covered by ext_authz by default
- Add a knob to enable ext_authz on direct response actions
- Add a knob to enable ext_authz on redirect actions

This way, this "bugfix" is now instead a new feature that won't break existing users that expect the current behavior. Principal of least surprise is violated, in theory, but we can mitigate that with documentation and I think this is better than possibly introducing a breaking change when we meant to make things better.

**Review by dio — APPROVED** [review]
Thanks! Could you update the PR description, so it will be easier for maintainers when landing this.

**Review by dio — APPROVED** [review]
Thanks! Could you update the PR description, so it will be easier for maintainers when landing this.

**Review by dio — APPROVED** [review]
Thanks! Could you update the PR description, so it will be easier for maintainers when landing this.

**Comment by dio** [comment]
Wow. GitHub put 3 approvals 😅. Sorry for that.

**Comment by esmet** [comment]
@dio thanks for the approval. But, I'm moving this back to draft because I have concerns that this bugfix could be considered a minor breaking change and also doesn't cover redirect responses (see https://github.com/envoyproxy/envoy/pull/14989#discussion_r577035720). When I come back to working on this PR, I'll look into either updating the changelog to mention that this is a behavioral change or add config knobs to control whether direct / redirect responses are covered by ext_authz by default.

**Comment by dio** [comment]
Oops. OK. Sorry for missing that comment.

**Comment by artificial-aidan** [comment]
> @dio thanks for the approval. But, I'm moving this back to draft because I have concerns that this bugfix could be considered a minor breaking change and also doesn't cover redirect responses (see [#14989 (comment)](https://github.com/envoyproxy/envoy/pull/14989#discussion_r577035720)). When I come back to working on this PR, I'll look into either updating the changelog to mention that this is a behavioral change or add config knobs to control whether direct / redirect responses are covered by ext_authz by default.

I thought about this for a bit. Would it be odd behavior to have redirect/direct responses default differently than other routes. Right now you need to disable a filter on a route otherwise it automatically applies. It seems like this has to be a breaking change, otherwise you end up with an unintuitive UI. :/

**Comment by esmet** [comment]
@artificial-aidan I agree that the user experience will be surprising (in a bad way) unless we break the current behavior.

Can I get an expert opinion from @envoyproxy/api-shepherds on whether the behavioral bugfix in this ticket / PR is small enough to simply be included in "Minor behavioral changes" in the changelog, or if it is big enough to deserve its own API switch?

The more I think on it, the more I'm leaning on noting it as a minor behavioral change because the current behavior is indeed an unintentional bug and there is a clear "workaround" to achieve the current behavior (by setting per filter config, as one would usually expect).

File: source/extensions/filters/http/ext_authz/ext_authz.cc (hunk 1 above)
Comment (esmet) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:385]: By the way, I was wrong. Redirects and direct replies are both "direct responses" on the route object so I believe this change will correctly apply to them all. I should probably add an integration test.

**Comment by htuch** [comment]
@esmet not sure if I'm seeing the API concern here. Could you elaborate on how this might impact someone operating an ext_authz service or what the change in semantics are?

If we're just saying "we fixed this to work for direct response entries as anyone who read the xDS docs would have expected anyway", I think we are good.

**Comment by esmet** [comment]
@htuch The question is essentially whether we think fixing this bug is enough of a behavioral change to actually be considered a breaking change, so I was looking to get an opinion on whether that meant we needed to add an API flag for this. There may have been some precedent I was unaware of, for example.

But I think you're saying that as long as this is a bug fix and the fixed behavior is reasonable to expect given the docs, we shouldn't need to do that. I agree and I'm OK with that too. I'll keep this PR as is in that case 👍

**Comment by artificial-aidan** [comment]
This would still be very helpful for us, and would simplify a lot of configs.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -27,6 +27,7 @@ Minor Behavior Changes
   logging, :ref:`auto_host_rewrite <envoy_api_field_route.RouteAction.auto_host_rewrite>`, etc.
   Setting the hostname manually allows overriding the internal hostname used for such features while
   still allowing the original DNS resolution name to be used.
+* ext_authz: routes that contain a redirect or a direct response action are now properly subject to external authorization checks by default. To restore the original behavior and skip authorization checks, disable ext_authz on the :ref:`per-route config <envoy_v3_api_msg_extensions.filters.http.ext_authz.v3.ExtAuthzPerRoute>`.
```
Comment (junr03) [inline docs/root/version_history/current.rst:30]: Should we call this out in the ext_authz docs?

File: source/extensions/filters/http/ext_authz/ext_authz.h [hunk 3]
```diff
@@ -268,6 +268,10 @@ class Filter : public Logger::Loggable<Logger::Id::filter>,
   struct PerRouteFlags {
     const bool skip_check_;
     const bool skip_request_body_buffering_;
+
+    static constexpr PerRouteFlags defaultFlags() { return PerRouteFlags{false, false}; }
```
Comment (junr03) [inline source/extensions/filters/http/ext_authz/ext_authz.h:272]: can we add comments on what the bools are here and below?

e.g., `false /*skip_check_*/, false /*skip_request_body_buffering_*/`

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 4]
```diff
@@ -1599,9 +1599,42 @@ TEST_P(HttpFilterTestParam, DisabledOnRouteWithRequestBody) {
...
+  EXPECT_CALL(*client_, check(_, _, _, _));
+  EXPECT_CALL(*filter_callbacks_.route_, routeEntry()).WillRepeatedly(Return(&route_entry));
+  EXPECT_EQ(Http::FilterHeadersStatus::StopAllIterationAndWatermark,
+            filter_->decodeHeaders(request_headers_, true));
+}
+
+// Test that the request stops at decodeHeaders when filter_callbacks has no route entry, but it
+// does have a direct response entry.
+TEST_P(HttpFilterTestParam, StopIterationWithDirectResponseEntry) {
+  NiceMock<Router::MockDirectResponseEntry> direct_response_entry;
```
Comment (junr03) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:1615]: should we have a test with redirect?

**Review by junr03 — COMMENTED** [review]
Thanks! Almost done

/wait

**Comment by artificial-aidan** [comment]
Still hoping to see this get in.

**Comment by artificial-aidan** [comment]
Not stale

**Comment by soulxu** [comment]
@artificial-aidan This should be already fixed by https://github.com/envoyproxy/envoy/pull/17449

**Comment by esmet** [comment]
@soulxu thanks for picking this up. I let a bunch of my work go stale unfortunately, this included.

**Pull request closed without merging**